### PR TITLE
[video_player] isCompleted event.

### PR DESCRIPTION
**Pull request by tarrinneal**
Adds `isCompleted` event to `VideoPlayerEvent`.

fixes https://github.com/flutter/flutter/issues/21929

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/video_player/video_player/lib/video_player.dart [hunk 1]
```diff
@@ -737,6 +756,7 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
     value = value.copyWith(
       position: position,
       caption: _getCaptionAt(position),
+      isCompleted: false,
```
Comment (tarrinneal) [inline packages/video_player/video_player/lib/video_player.dart:759]: I'm not sure whether this is needed/wanted. I do worry that it may cause inconsistencies since the logic that sets `isCompleted` to true also seeks to the end.

File: packages/video_player/video_player/lib/video_player.dart [hunk 2]
```diff
@@ -528,7 +547,7 @@ class VideoPlayerController extends ValueNotifier<VideoPlayerValue> {
   /// Sets whether or not the video should loop after playing once. See also
   /// [VideoPlayerValue.isLooping].
   Future<void> setLooping(bool looping) async {
-    value = value.copyWith(isLooping: looping);
+    value = value.copyWith(isLooping: looping, isCompleted: false);
```
Comment (stuartmorgan) [inline packages/video_player/video_player/lib/video_player.dart:550]: Does enabling looping start playback? I would have assumed not, and if it's sitting paused and completed, enabling looping for the future doesn't seem like it should clean `isCompleted`.

File: packages/video_player/video_player/lib/video_player.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/video_player/video_player/lib/video_player.dart:759]: Can't we pass nil here if `position == duration` to avoid that logic clearing the flag?

File: packages/video_player/video_player/test/video_player_test.dart [hunk 3]
```diff
@@ -1204,6 +1205,64 @@ void main() {
...
+        if (controller.value.isCompleted) {
+          isCompletedTest();
+        }
+      }
+    });
+
+    fakeVideoEventStream.add(VideoEvent(eventType: VideoEventType.completed));
+  });
+
+  test('isCompleted updates on video end', () async {
```
Comment (stuartmorgan) [inline packages/video_player/video_player/test/video_player_test.dart:1236]: This has the same name as the test above.

File: packages/video_player/video_player/test/video_player_test.dart [hunk 4]
```diff
@@ -1204,6 +1205,64 @@ void main() {
...
+              .add(VideoEvent(eventType: VideoEventType.isPlayingStateUpdate));
+        } else {
+          isNoLongerCompletedTest();
+        }
+      }
+    });
+
+    fakeVideoEventStream.add(VideoEvent(eventType: VideoEventType.completed));
+  });
 }
```
Comment (stuartmorgan) [inline packages/video_player/video_player/test/video_player_test.dart:1267]: There should be a test for each case that is supposed to set `isCompleted` back to `false`.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/video_player/video_player/lib/video_player.dart (hunk 1 above)
Comment (tarrinneal) [reply packages/video_player/video_player/lib/video_player.dart:759]: I set this to just be set to position == duration.

File: packages/video_player/video_player/test/video_player_test.dart (hunk 3 above)
Comment (tarrinneal) [reply packages/video_player/video_player/test/video_player_test.dart:1236]: oops

**Comment by tarrinneal** [comment]
> LGTM
> 
> I suspect we'll need to tune edge cases for this state based on feedback from real-world use, but this looks good as a baseline for that.

Agreed, I'll keep an eye on it.
